src/mlpack/methods/linear_svm/linear_svm_main.cpp

```diff
@@ -0,0 +1,440 @@
+  if (CLI::HasParam("training") && CLI::HasParam("labels"))
+  {
+    rawLabels = std::move(CLI::GetParam<arma::Row<size_t>>("labels"));
+    if (trainingSet.n_cols != rawLabels.n_cols)
+    {
+      Log::Fatal << "The labels must have the same number of points as the "
+          << "training dataset." << endl;
+    }
+  }
+  else if (CLI::HasParam("training"))
+  {
+    // Checking the size of training data if no labels are passed.
+    if (trainingSet.n_rows < 2)
+    {
+      Log::Fatal << "Can't get labels from training data since it has less "
+          << "than 2 rows." << endl;
+    }
+
+    // The initial predictors for y, Nx1.
+    rawLabels = arma::conv_to<arma::Row<size_t>>::from(
+        trainingSet.row(trainingSet.n_rows - 1));
+    trainingSet.shed_row(trainingSet.n_rows - 1);
+  }
+
+  // Load the model, if necessary.
+  LinearSVMModel* model;
+  if (CLI::HasParam("input_model"))
+  {
+    model = CLI::GetParam<LinearSVMModel*>("input_model");
+  }
+  else
+  {
+    model = new LinearSVMModel();
+  }
+
+  // Now, do the training.
+  if (CLI::HasParam("training"))
+  {
+    data::NormalizeLabels(rawLabels, labels, model->mappings);
+    numClasses = CLI::GetParam<int>("num_classes") == 0 ?
+        arma::max(labels) + 1 : CLI::GetParam<int>("num_classes");
+    model->svm.Lambda() = lambda;
+    model->svm.Delta() = delta;
+    model->svm.NumClasses() = numClasses;
+
+    if (optimizerType == "lbfgs")
+    {
+      model->svm.FitIntercept() = intercept;
+
+      ens::L_BFGS lbfgsOpt;
+      lbfgsOpt.MaxIterations() = maxIterations;
+      lbfgsOpt.MinGradientNorm() = tolerance;
+
+      Log::Info << "Training model with L-BFGS optimizer." << endl;
+
+      // This will train the model.
+      model->svm.Train(trainingSet, labels, numClasses, lbfgsOpt);
+    }
+    else if (optimizerType == "psgd")
+    {
+      const double stepSize = CLI::GetParam<double>("step_size");
+      const bool shuffle = !CLI::HasParam("shuffle");
+      const size_t maxIt = epochs * trainingSet.n_cols;
+
+      ens::ConstantStep decayPolicy(stepSize);
+
+      #ifdef HAS_OPENMP
+      size_t threads = omp_get_max_threads();
+      #else
+      size_t threads = 1;
+      #endif
+
+      ens::ParallelSGD<ens::ConstantStep> psgdOpt(maxIt, std::ceil(
+        (float) trainingSet.n_cols / threads), tolerance, shuffle,
+        decayPolicy);
+
+      Log::Info << "Training model with ParallelSGD optimizer." << endl;
+
+      // This will train the model.
+      model->svm.Train(trainingSet, labels, numClasses, psgdOpt);
+    }
+  }
+  if (CLI::HasParam("test"))
+  {
+    if (!CLI::HasParam("training"))
+    {
+      numClasses = model->svm.NumClasses();
+    }
+    // Get the test dataset, and get predictions.
+    testSet = std::move(CLI::GetParam<arma::mat>("test"));
+    arma::Row<size_t> predictions;
+    size_t trainingDimensionality;
+
+    // Set the dimensionality according to fitintercept.
+    if (intercept && optimizerType == "lbfgs")
```

Comment:
Why is this specific to L-BFGS being used?  Shouldn't we be able to fit the intercept for PSGD also?

Reply:
Yes this is strange but here we get inf and zero for prediction and scores when we use fit the intercept for psgd.

Reply:
I see.  What happens when only one thread is used when we train with PSGD and fit the intercept?  I think I understand why this is happening, but need to know what happens in the singlethreaded case before I am sure.